Add k-NN Faiss filtering documentation (#4476)
opensearch-project/documentation-website: _search-plugins/knn/filter-search-knn.md

To refine k-NN results, you can filter a k-NN search using one of the following methods:

> - [Efficient k-NN filtering](#efficient-k-nn-filtering): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned. This approach is supported by the following search engines:

Reviewer: ```suggestion
- [Efficient k-NN filtering](#efficient-k-nn-filtering): This approach applies filtering _during_ the k-NN search, as opposed to before or after the k-NN search, which ensures that `k` results are returned(if present). This approach is supported by the following engines:
```

Author: Reworded.

>   - Lucene search engine with a Hierarchical Navigable Small World (HNSW) algorithm (k-NN plugin versions 2.4 and later) 

Reviewer: ```suggestion
  - Lucene engine with a Hierarchical Navigable Small World (HNSW) algorithm (k-NN plugin versions 2.4 and later) 
```

>   - Faiss search engine with a Hierarchical Navigable Small World (HNSW) algorithm (k-NN plugin versions 2.9 or later) 

Reviewer: ```suggestion
  - Faiss engine with a Hierarchical Navigable Small World (HNSW) algorithm (k-NN plugin versions 2.9 or later) 
```

>   - Faiss engine with a Hierarchical Navigable Small World (HNSW) algorithm (k-NN plugin versions 2.9 or later) 

Reviewer 2: ```suggestion
  - Faiss engine with a HNSW algorithm (k-NN plugin versions 2.9 or later) 
```

> -  [Post filtering](#post-filtering): Because it is performed after the k-NN search, this approach may return significantly fewer than `k` results for a restrictive filter.

Reviewer 3: We hyphenate "post-filtering" later in the doc.

>     - [Boolean post filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search and then applies a filter to the results. The two query parts are executed independently and then the intersection of their result sets is taken. 

Reviewer: ```suggestion
    - [Boolean post filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search and then applies a filter to the results. The two query parts are executed independently and then the results are combined based on the query operator(should/must etc) provided in the query.
```

>     - [Boolean post filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search and then applies a filter to the results. The two query parts are executed independently and then the results are combined based on the query operator (`should`, `must`, and so on) provided in the query. 

Reviewer 3: ```suggestion
    - [Boolean post filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search and then applies a filter to the results. The two query parts are executed independently, and then the results are combined based on the query operator (`should`, `must`, and so on) provided in the query. 
```

>     - [Boolean post filter](#boolean-filter-with-ann-search): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search and then applies a filter to the results. The two query parts are executed independently and then the results are combined based on the query operator (`should`, `must`, and so on) provided in the query. 

Reviewer 3: "Boolean post-filter"? "elements" instead of "parts"? "run" instead of "executed"?

>     - [The `post_filter` parameter](#post-filter-parameter): This approach runs an [ANN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) on the full dataset and then applies the filter to k-NN results.

Reviewer 3: ```suggestion
    - [The `post_filter` parameter](#post-filter-parameter): This approach runs an [ANN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) search on the full dataset and then applies the filter to k-NN results.
```

>     - [The `post_filter` parameter](#post-filter-parameter): This approach runs an [ANN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) on the full dataset and then applies the filter to k-NN results.

Reviewer 3: Should "the" precede "k-NN results"?

>     - [The `post_filter` parameter](#post-filter-parameter): This approach runs an [approximate nearest neighbor (ANN)]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) on the full dataset and then applies the filter to k-NN results.

Reviewer 2: ```suggestion
    - [The `post_filter` parameter](#post-filter-parameter): This approach runs an [ANN]({{site.url}}{{site.baseurl}}/search-plugins/knn/approximate-knn/) on the full dataset and then applies the filter to k-NN results.
```

> - [Scoring script filter](#scoring-script-filter): This approach involves pre-filtering a document set and then running an exact k-NN search on the filtered subset. It does not scale for large filtered subsets. 

Reviewer: Lets also add that, latencies can be high for this query.

> - [Scoring script filter](#scoring-script-filter): This approach involves pre-filtering a document set and then running an exact k-NN search on the filtered subset. It may have high latency and does not scale for large filtered subsets. 

Reviewer 3: Instead of "for", can this be something like "does not scale to accommodate large filtered subsets"?

The following table summarizes the preceding filtering use cases.

Filter | When the filter is applied | Type of search | Supported engines and methods | Where to place the `filter` clause
:--- | :--- | :--- | :---

> Efficient k-NN filtering | During search (a hybrid of pre- and post-filtering) | Approximate | - `lucene` (`hnsw`) <br> - `faiss` (`hnsw`, `ivf`) | Inside the k-NN query clause.

Reviewer: lets remove ivf from here.

Once you've estimated the number of documents in your index, the restrictiveness of your filter, and the desired number of nearest neighbors, use the following table to choose a filtering method that optimizes for recall or latency.

<!-- TODO review this table given the new faiss engine -->

| Number of documents in an index | Percentage of documents the filter returns | k | Filtering method to use for higher recall | Filtering method to use for lower latency |
| :-- | :-- | :-- | :-- | :-- |
| 10M | 2.5 | 100 | Scoring script | Scoring script |
| 10M | 38 | 100 |Efficient k-NN filtering | Boolean filter |
| 10M | 80 | 100 | Scoring script |Efficient k-NN filtering |
| 1M | 2.5 | 100 |Efficient k-NN filtering | Scoring script |
| 1M | 38 | 100 |Efficient k-NN filtering |Efficient k-NN filtering/scoring script |

> | 1M | 80 | 100 | Boolean filter |Efficient k-NN filtering |

Reviewer: ```suggestion
| 1M | 80 | 100 | Efficient k-NN filtering |Boolean filter |
```

## Efficient k-NN filtering

> You can perform efficient k-NN filtering with the `lucene` or `faiss` search engines. 

Reviewer: ```suggestion
You can perform efficient k-NN filtering with the `lucene` or `faiss` engines. 
```

}
```

For more ways to construct a filter, see [Constructing a filter](#constructing-a-filter).

### Faiss k-NN filter implementation 

Starting with k-NN plugin version 2.9, you can use `faiss` filters for k-NN searches.

When you specify a Faiss filter for a k-NN search, the Faiss algorithm decides whether to perform an exact k-NN search with pre-filtering or an approximate search with modified post-filtering. The algorithm uses the following variables:

- N: The number of documents in the index.
- P: The number of documents in the document subset after the filter is applied (P <= N).
- k: The maximum number of vectors to return in the response.

The following flow chart outlines the Faiss algorithm.

![Faiss algorithm for filtering]({{site.url}}{{site.baseurl}}/images/faiss-algorithm.jpg)

### Using a Faiss efficient filter

> Consider an index that holds information about shirts for a clothing store. You want to find the top rated shirts that are similar to the one you have but would like to restrict the results by shirt size.

Reviewer 2: ```suggestion
Consider an index that holds information about shirts for a clothing store. You want to find the top-rated shirts that are similar to the one you have but would like to restrict the results by shirt size.
```

> Consider an index that holds information about shirts for a clothing store. You want to find the top-rated shirts that are similar to the one you have but would like to restrict the results by shirt size.

Reviewer 3: ```suggestion
Consider an index containing information about a particular brand of shirt. You want to find the top-rated shirts that are similar to one you already have but would like to restrict the results by shirt size.
```

  }
} 
```
{% include copy-curl.html %}

### Faiss k-NN filter implementation 

Starting with k-NN plugin version 2.9, you can use `faiss` filters for k-NN searches.

<!-- TODO: add a description and diagram of the new algorithm -->

### Using a Faiss efficient filter

<!-- TODO: describe a Faiss example -->

**Step 1: Create a new index** 

Before you can run a k-NN search with a filter, you need to create an index with a `knn_vector` field. For this field, you need to specify `faiss` as the engine in the mapping.

The following request <!-- TODO: what does the request do -->:

```json
<!-- TODO: add a request -->
```
{% include copy-curl.html %}

**Step 2: Add data to your index**

Next, add data to your index.

The following request <!-- TODO: what does the request do -->:  

```json
<!-- TODO: add a request -->
```
{% include copy-curl.html %}

**Step 3: Search your data with a filter**

> Now you can create a k-NN search with filters. <!-- TODO: add details -->

Reviewer: The same thing which we wrote for lucene above is true for filters here too.

```
Note that there are multiple ways to construct a filter that returns hotels that provide parking, for example:

A term query clause in the should clause
A wildcard query clause in the should clause
A regexp query clause in the should clause
A must_not clause to eliminate hotels with parking set to false.
```

  }
} 
```
{% include copy-curl.html %}

> ## Post filtering

Reviewer 3: This is hyphenated elsewhere. Ensure consistency across docs.

> You can achieve post filtering with a Boolean filter or by providing the `post_filter` parameter.

Reviewer 3: Same

            }
          }
        }
      ]
    }
  }
}
```

The response includes documents containing the matching hotels:

```json
{
  "took" : 95,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 5,
      "relation" : "eq"
    },
    "max_score" : 0.72992706,
    "hits" : [
      {
        "_index" : "hotels-index",
        "_id" : "3",
        "_score" : 0.72992706,
        "_source" : {
          "location" : [
            4.9,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "6",
        "_score" : 0.3012048,
        "_source" : {
          "location" : [
            6.4,
            3.4
          ],
          "parking" : "true",
          "rating" : 9
        }
      },
      {
        "_index" : "hotels-index",
        "_id" : "5",
        "_score" : 0.24154587,
        "_source" : {
          "location" : [
            3.3,
            4.5
          ],
          "parking" : "true",
          "rating" : 8
        }
      }
    ]
  }
}
```

> ### Post filter parameter

Reviewer 3: "Post-filter?

> If you use the `knn` query alongside filters or other clauses (e.g. `bool`, `must`, `match`), you might receive fewer than `k` results. In this example, `post_filter` reduces the number of results from 2 to 1:

Reviewer 2: ```suggestion
If you use the `knn` query alongside filters or other clauses (for example, `bool`, `must`, `match`), you might receive fewer than `k` results. In this example, `post_filter` reduces the number of results from 2 to 1:
```

```json
GET my-knn-index-1/_search
{
  "size": 2,
  "query": {
    "knn": {
      "my_vector2": {
        "vector": [2, 3, 5, 6],
        "k": 2
      }
    }
  },
  "post_filter": {
    "range": {
      "price": {
        "gte": 5,
        "lte": 10
      }
    }
  }
}
```

## Scoring script filter

> A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the 3 hotels that are closest to the specified `location`:

Reviewer 2: ```suggestion
A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the three hotels that are closest to the specified `location`:
```

> A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the three hotels that are closest to the specified `location`:

Reviewer 3: ```suggestion
A scoring script filter first filters the documents and then uses a brute-force exact k-NN search on the results. For example, the following query searches for hotels with a rating between 8 and 10, inclusive, that provide parking and then performs a k-NN search to return the 3 hotels that are closest to the specified `location`:
```
